### PR TITLE
scheduled sampling with example on voxforge asr1

### PR DESCRIPTION
**Pull request by creatorscan**
@sw005320 @karaf 
Implemented the following changes:

- Having scheduled sampling in Decoder function

- A modified run.sh script with RESULTS for voxforge database

### CODE REVIEW

File: src/nets/e2e_asr_attctc_th.py [hunk 1]
```diff
@@ -356,6 +357,7 @@ def recognize(self, x, recog_args, char_list, rnnlm=None):
 
         # 1. encoder
         # make a utt list (1) to use the same interface for encoder
+        h = h.contiguous()
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc_th.py]: Is it required?

File: egs/voxforge/asr1/run.sh [hunk 2]
```diff
@@ -171,7 +174,7 @@ if [ ${stage} -le 2 ]; then
 fi
 
 if [ -z ${tag} ]; then
-    expdir=exp/${train_set}_${etype}_e${elayers}_subsample${subsample}_unit${eunits}_proj${eprojs}_d${dlayers}_unit${dunits}_${atype}_aconvc${aconv_chans}_aconvf${aconv_filts}_mtlalpha${mtlalpha}_${opt}_bs${batchsize}_mli${maxlen_in}_mlo${maxlen_out}
+    expdir=exp/${train_set}_${etype}_e${elayers}_subsample${subsample}_unit${eunits}_proj${eprojs}_d${dlayers}_unit${dunits}_${atype}_aconvc${aconv_chans}_aconvf${aconv_filts}_mtlalpha${mtlalpha}_${opt}_sampratio${samp_ratio}_bs${batchsize}_mli${maxlen_in}_mlo${maxlen_out}_epochs${epochs}
```
Comment (sw005320) [inline egs/voxforge/asr1/run.sh]: can you remove `_epochs${epochs}`?

File: egs/voxforge/asr1/run.sh [hunk 3]
```diff
@@ -55,6 +55,9 @@ minlenratio=0.0
 ctc_weight=0.3
 recog_model=acc.best # set a model to be used for decoding: 'acc.best' or 'loss.best'
 
+# scheduled sampling option
```
Comment (sw005320) [inline egs/voxforge/asr1/run.sh]: Add these scheduled sampling related changes to all ASR recipes.

File: egs/voxforge/asr1/run.sh [hunk 4]
```diff
@@ -210,6 +213,7 @@ if [ ${stage} -le 3 ]; then
         --maxlen-in ${maxlen_in} \
         --maxlen-out ${maxlen_out} \
         --opt ${opt} \
+        --scheduled-sampling-ratio ${samp_ratio} \
```
Comment (sw005320) [inline egs/voxforge/asr1/run.sh]: The option name would be a bit too long.
Any idea?

File: src/nets/e2e_asr_attctc_th.py (hunk 1 above)
Comment (creatorscan) [reply src/nets/e2e_asr_attctc_th.py]: This is required when we use VGG layers initially..

File: egs/voxforge/asr1/run.sh (hunk 2 above)
Comment (creatorscan) [reply egs/voxforge/asr1/run.sh]: ok Shinji..

File: egs/voxforge/asr1/run.sh (hunk 4 above)
Comment (creatorscan) [reply egs/voxforge/asr1/run.sh]: What about sample-probability ?

Comment (creatorscan) [reply egs/voxforge/asr1/run.sh]: I will keep it as sampling-probability, as it sounds good..

File: egs/wsj/asr1/run.sh [hunk 5]
```diff
@@ -7,7 +7,7 @@
 . ./cmd.sh
 
 # general configuration
-backend=chainer
+backend=pytorch
```
Comment (sw005320) [inline egs/wsj/asr1/run.sh]: Can you keep it as it is?

File: egs/wsj/asr1/run.sh [hunk 6]
```diff
@@ -68,9 +68,14 @@ minlenratio=0.0
 ctc_weight=0.3
 recog_model=acc.best # set a model to be used for decoding: 'acc.best' or 'loss.best'
 
+# scheduled sampling option
+samp_prob=0.0
+
 # data
 wsj0=/export/corpora5/LDC/LDC93S6B
 wsj1=/export/corpora5/LDC/LDC94S13B
+wsj0=/mnt/matylda2/data/WSJ/WSJ0
```
Comment (sw005320) [inline egs/wsj/asr1/run.sh]: Can you keep it as it is?

File: egs/wsj/asr1/run.sh [hunk 7]
```diff
@@ -192,8 +197,8 @@ else
     lmdict=$dict
 fi
 mkdir -p ${lmexpdir}
-if [ ${stage} -le 3 ]; then
-    echo "stage 3: LM Preparation"
+if [ ${stage} -le 4 ]; then
```
Comment (sw005320) [inline egs/wsj/asr1/run.sh]: Can you keep it as it is?

File: egs/wsj/asr1/run.sh [hunk 8]
```diff
@@ -242,8 +247,8 @@ else
 fi
 mkdir -p ${expdir}
 
-if [ ${stage} -le 4 ]; then
-    echo "stage 4: Network Training"
+if [ ${stage} -le 3 ]; then
```
Comment (sw005320) [inline egs/wsj/asr1/run.sh]: Can you keep it as it is?

**Comment by sw005320** [comment]
Also, consider
> 1. add a test script
> 2. make this PR to the dev branch. I want to use this function in v.0.2.0.

**Comment by sw005320** [comment]
Thanks.
Can you add either
- add the same function to the chainer backend
or
- add some error handling in the chainer backend when this option is specified, similar to https://github.com/espnet/espnet/blob/master/src/asr/asr_chainer.py#L294-L296 (if fails if sampling_probability > 0)

**Comment by sw005320** [comment]
@creatorscan, can you try to finish this PR by fixing the following items? 
- There are conflicts mainly come from the change in v.0.2.0. You could continue to work on this PR or #358, and please fix them.
- As @hirofumi0810 pointed out, you may not need to perform softmax, which would save the computational cost. 
- Again, @hirofumi0810 told me that it may not work when the number of decoder layers is more than 1.

**Comment by creatorscan** [comment]
By including softmax, there was small gains observed during experimental analysis in all databases. The results shows the performance of eval set in voxforge with and without softmax.
|  Type           |  % CER | %WER  |
|----------------|----------------|--------|
|  Schedule sampling     |      15.4      | 50.9 |
|  +softmax        |         14.8        | 50.4 |
So, that was the reason to have the softmax in the code. Currently, I dont have much intuition to justify this scenario.
